### PR TITLE
Several optimizations

### PR DESCRIPTION
**Pull request by fatkodima**
Relevant - https://github.com/rubocop/rubocop/pull/11279.

### CODE REVIEW

File: lib/rubocop/ast/node/mixin/method_dispatch_node.rb [hunk 1]
```diff
@@ -211,7 +211,7 @@ def lambda?
       #
       # @return [Boolean] whether this method is a lambda literal
       def lambda_literal?
-        block_literal? && loc.expression && loc.expression.source == '->'
+        loc.expression && loc.expression.source == '->' && block_literal?
```
Comment (iliabylich) [inline lib/rubocop/ast/node/mixin/method_dispatch_node.rb:214]: are there nodes that don't have `.loc.expression`?

File: lib/rubocop/ast/node.rb [hunk 2]
```diff
@@ -101,11 +105,19 @@ def initialize(type, children = [], properties = {})
...
         method_name = "#{node_type.to_s.gsub(/\W/, '')}_type?"
-        define_method(method_name) do
-          type == node_type
-        end
+        class_eval <<~RUBY, __FILE__, __LINE__ + 1
+          def #{method_name}          # def block_type?
+            @type == :#{node_type}    #   @type == :block
+          end                         # end
+        RUBY
+      end
```
Comment (fatkodima) [inline lib/rubocop/ast/node.rb:115]: This code is in the hot path. Executing `class_eval`-defined method is 60% faster than by `define_method`.
Comparing instance variable is 20% faster that calling an `attr_accessor`-defined method.

File: lib/rubocop/ast/node.rb [hunk 3]
```diff
@@ -101,11 +105,19 @@ def initialize(type, children = [], properties = {})
...
+
+      # Most nodes are of 'send' type, so this method is defined
+      # separately to make this check as fast as possible.
+      def send_type?
+        false
       end
```
Comment (fatkodima) [inline lib/rubocop/ast/node.rb:121]: Most checks are done against `send` nodes, so I extracted a dedicated fastest method for this.

File: lib/rubocop/ast/node/mixin/method_dispatch_node.rb (hunk 1 above)
Comment (fatkodima) [reply lib/rubocop/ast/node/mixin/method_dispatch_node.rb]: Seems like no. Thank you, removed that check.

**Comment by marcandre** [comment]
Looks great!
I thought I had merged it, sorry for the delay.

**Comment by marcandre** [comment]
Released in 1.14.1

**Pull request merged**